File: Medium/198-House-Robber.java
// Time: O(n), 1D DP

class Solution {
    public int rob(int[] nums) {
        int len = nums.length;
        if (len == 1) {
            return nums[0];
        }
        
        int[] dp = new int[len];
        dp[len - 1] = nums[len - 1];
        dp[len - 2] = Math.max(nums[len - 1], nums[len - 2]);
        
        for (int i = len - 3; i >= 0; i--) {
            dp[i] = Math.max(dp[i + 1], dp[i + 2] + nums[i]);
        }
             
        return dp[0];
    }
}

File: Medium/198-House-Robber.py
class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if len(nums) == 1:
            return nums[0]

        a, b = nums[0], max(nums[0], nums[1])

        for i in range(2, len(nums)):
            tmp = b
            b = max(b, nums[i] + a)
            a = tmp
        
        return max(a, b)

        
